src/bootstrap/src/utils/build_stamp.rs

```diff
@@ -0,0 +1,194 @@
+//! Module for managing build stamp files.
+//!
+//! Contains the core implementation of how bootstrap utilizes stamp files on build processes.
+
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+
+use sha2::digest::Digest;
+
+use crate::core::builder::Builder;
+use crate::core::config::TargetSelection;
+use crate::utils::helpers::{hex_encode, mtime};
+use crate::{Compiler, Mode, helpers, t};
+
+#[cfg(test)]
+mod tests;
+
+/// Manages a stamp file to track build state. The file is created in the given
+/// directory and can have custom content and name.
+#[derive(Clone)]
+pub struct BuildStamp {
+    path: PathBuf,
+    pub(crate) stamp: String,
+}
+
+impl AsRef<Path> for BuildStamp {
+    fn as_ref(&self) -> &Path {
+        &self.path
+    }
+}
+
+impl BuildStamp {
+    /// Creates a new `BuildStamp` for a given directory.
+    ///
+    /// By default, stamp will be an empty file named `.stamp` within the specified directory.
+    pub fn new(dir: &Path) -> Self {
+        // Avoid using `is_dir()` as the directory may not exist yet.
+        // It is more appropriate to assert that the path is not a file.
+        assert!(!dir.is_file(), "can't be a file path");
```

Comment:
Question: does it matter if `dir` is a symbolic link?

Reply:
It does not. See the [doc](https://doc.rust-lang.org/std/path/struct.Path.html#method.is_file).